[Feature] - Course Management Detail View

FuchsDominik opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
The course management detail view should be integrated more deeply into the workflow of the instructors and tutors. Currently the detail page is not used very frequently since it only stores basic course information which is not useful for daily tasks

### Description
<!-- Describe your changes in detail -->
In the first iteration of this page, we add an active user chart and some doughnut charts to visualize some statistics regarding the course

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate into Course Management Overview
3. Click on Course header
4. In the Course Management Detail View, make sure that no weird numbers show up which are not rounded for example
5. Create a new exercise and participate with min. 2 users
6. Check in the detail view that two more assessments need to be done
7. Assess those and create a complaint with the one and a more feedback request with the other one
8. Check in the detail view that the corresponding numbers in the doughnut charts go up
9. Check that the active users rise by the amount of user with which you made submissions
10. Check that the average score of the course is correct (average student score is shown in the course statistics and in the participant score)

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->
course-detail.component.ts: 93%
course-detail-bar-chart.component.ts: 95%
course-detail-doughnut-chart.component.ts: 89%
CourseResource.java: 93%

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
Example of a course detail page
![image](https://user-images.githubusercontent.com/44401560/116396421-a7f46480-a825-11eb-9d8b-14e018e32df3.png)

We also still provide the old course data below the charts:
<img width="1436" alt="Screenshot 2021-05-04 at 13 48 20" src="https://user-images.githubusercontent.com/44401560/116999345-c2c05080-acdf-11eb-8226-1d9b416a6d27.png">

In comparison, before, it looked like this:
![Screenshot 2021-05-04 at 13 51 52](https://user-images.githubusercontent.com/44401560/116999473-f4391c00-acdf-11eb-80f7-b06b4d277e61.png)

willergeorg reviewed (changes requested):
In general works good. I just had a problem regarding the average student score. There it says 0% and 0/0 whereas the participation scores and the statistic say something about 33.4% (see screenshots below).


![issue8](https://user-images.githubusercontent.com/80622272/116229211-824c5a00-a756-11eb-9918-b0da70866502.PNG)
![issue9](https://user-images.githubusercontent.com/80622272/116229224-87110e00-a756-11eb-8691-f4bff35bf549.PNG)

FuchsDominik commented:
> In general works good. I just had a problem regarding the average student score. There it says 0% and 0/0 whereas the participation scores and the statistic say something about 33.4% (see screenshots below).
> 
> ![issue8](https://user-images.githubusercontent.com/80622272/116229211-824c5a00-a756-11eb-9918-b0da70866502.PNG)
> ![issue9](https://user-images.githubusercontent.com/80622272/116229224-87110e00-a756-11eb-8691-f4bff35bf549.PNG)

Thanks, fixed your edge case in 0b36195

willergeorg reviewed (approved):
re-approve. Issue fixed.

NicolasUllmann reviewed (dismissed):
Tested on Server 1. Everything worked as expected.

katjanakosic reviewed (dismissed):
Tested on TS1. Everything looks good.

maximilianbergmann reviewed (approved):
Tested on TS3. Works as expected

ArtemTsurikov reviewed (dismissed):
Tested on ts2. Works as expected now!

real-dmy reviewed (dismissed):
Tested on TS2. Works as expected.

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -303,15 +304,17 @@ public void registerUserForCourse(User user, Course course) {
         ZoneId zone = now.getZone();
-        ZonedDateTime startDate = localStartDate.atZone(zone).minusWeeks(3).withHour(0).withMinute(0).withSecond(0).withNano(0);
-        ZonedDateTime endDate = localEndDate.atZone(zone).withHour(23).withMinute(59).withSecond(59);
+        ZonedDateTime startDate = localStartDate.atZone(zone).minusWeeks(3 + (4 * (-periodIndex))).withHour(0).withMinute(0).withSecond(0).withNano(0);
+        ZonedDateTime endDate = periodIndex != 0 ? localEndDate.atZone(zone).minusWeeks(4 * (-periodIndex)).withHour(23).withMinute(59).withSecond(59)
+                : localEndDate.atZone(zone).withHour(23).withMinute(59).withSecond(59);
```
I'm still a bit confused here, especially the two calculations 3 + (4 * (-periodIndex)) and 4 * (-periodIndex). But I think this would make it a bit more intuitive:
```suggestion
        ZonedDateTime startDate = localStartDate.atZone(zone).minusWeeks(3 + (4 * (-periodIndex))).withHour(0).withMinute(0).withSecond(0).withNano(0);
        ZoneDateTime someFittingName = periodIndex != 0 ? localEndDate.atZone(zone).minusWeeks(4 * (-periodIndex)) : localEndDate;
        ZonedDateTime endDate = someFittingName.withHour(23).withMinute(59).withSecond(59);
```

FuchsDominik replied:
What should the `someFittingName` portray? In your case someFittingName could have two types, either ZonedDateTime if periodIndex != null or LocalDateTime if periodIndex == 0, if i am not mistaken. The reason why we distinguish for the endDate is that if we are looking at the current week, we take the end of the current day as the endDate. If we, however, look at another week, the endDate is the Sunday 23:59 of that particular week.

JohannesStoehr replied:
You're right, I forgot the .atZone() in the second part of the ternary.
The someFittingName would be the end date without the changed time. I mostly tried to make the code a bit more readable...

FuchsDominik replied:
I added comments which describe what is happening here in e9883b7

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -395,6 +398,26 @@ private Integer getWeekOfDate(ZonedDateTime date) {
+        var course = this.courseRepository.findByIdElseThrow(courseId);
+        dto.setCourse(course);
+
+        dto.setNumberOfStudentsInCourse(Math.toIntExact(userRepository.countUserInGroup(course.getStudentGroupName())));
+        dto.setNumberOfTeachingAssistantsInCourse(Math.toIntExact(userRepository.countUserInGroup(course.getTeachingAssistantGroupName())));
+        dto.setNumberOfInstructorsInCourse(Math.toIntExact(userRepository.countUserInGroup(course.getInstructorGroupName())));
```
You're missing the editor that got introduced yesterday.

FuchsDominik replied:
Thanks, forget to add the manual fetching, added in 36c3339 :)

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -1386,4 +1392,91 @@ private boolean checkIfUserIsMemberOfCourseOrganizations(User user, Course cours
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.TEACHING_ASSISTANT, course, null);
+
+        var exercises = exerciseRepository.findAllExercisesByCourseId(courseId);
+        var includedExercises = exercises.stream().filter(Exercise::isCourseExercise)
+                .filter(exercise -> !exercise.getIncludedInOverallScore().equals(IncludedInOverallScore.NOT_INCLUDED)).collect(Collectors.toSet());
+        var averageScoreForCourse = participantScoreRepository.findAvgScore(includedExercises);
+        averageScoreForCourse = averageScoreForCourse != null ? averageScoreForCourse : 0.0;
+        var reachablePoints = includedExercises.stream().map(Exercise::getMaxPoints).collect(Collectors.toSet()).stream().mapToDouble(Double::doubleValue).sum();
```
You use a lot of var here. At least I'd prefer real types here.

FuchsDominik replied:
We could have a little discussion about that in the next weekly meeting. I personally started using var a lot since working on this project because it is used very heavily here. If we decide to use less vars i can refactor this, of course

FuchsDominik replied:
Added it to the meeting discussion for next meeting

FuchsDominik replied:
Sounds good

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -1386,4 +1392,91 @@ private boolean checkIfUserIsMemberOfCourseOrganizations(User user, Course cours
+        // Average Student Score
+        dto.setCurrentMaxAverageScore(reachablePoints);
+        dto.setCurrentAbsoluteAverageScore(round((averageScoreForCourse / 100.0) * reachablePoints));
+        if (reachablePoints > 0.0) {
```
Is it possible that some edge cases with rounding occur in that line?

FuchsDominik replied:
We assure that neither reachablePoints nor averageScoreForCourse is null (Stream.sum() returns a `double` and I added a check for a possible average score null values earlier in this method). So i could not think of an scenario where this would make problems...

JohannesStoehr commented on src/main/webapp/app/course/manage/detail/course-detail.component.scss:
```diff
@@ -0,0 +1,53 @@
+jhi-secured-image {
+    padding-right: 20px;
+
+    ::ng-deep img {
```
Here are some Codacy issues. Are you aware of that?

FuchsDominik replied:
Yes but this was not changed, i just moved the class to another folder. Since css is often wizardry, i decided to leave if like that to not brake anything

JohannesStoehr commented on src/main/webapp/app/course/manage/detail/course-detail.component.scss:
```diff
@@ -0,0 +1,53 @@
+}
+
+.course-icon {
+    ::ng-deep {
```
One more Codacy error

FuchsDominik replied:
same as above

JohannesStoehr commented on src/test/java/de/tum/in/www1/artemis/util/CourseTestService.java:
```diff
@@ -1552,4 +1552,147 @@ public void testGetExerciseStatsForCourseOverviewWithPastExercises() throws Exce
+        ZonedDateTime now = ZonedDateTime.now();
+        // add courses with exercises
+
+        var course = database.createCoursesWithExercisesAndLectures(true).get(0);
+
+        var student1 = ModelFactory.generateActivatedUser("user1");
+        var student2 = ModelFactory.generateActivatedUser("user2");
+        // Fetch and update an instructor
+        var instructor = database.getUserByLogin("instructor1");
+        var groups = new HashSet<String>();
+        groups.add("instructor");
+        instructor.setGroups(groups);
+
+        userRepo.save(instructor);
+        userRepo.save(student1);
+        userRepo.save(student2);
+
+        var releaseDate = now.minusDays(7);
+        var dueDate = now.minusDays(2);
+        var assessmentDueDate = now.minusDays(1);
+        var exercise = ModelFactory.generateTextExercise(releaseDate, dueDate, assessmentDueDate, course);
+        exercise.setMaxPoints(5.0);
+        exercise = exerciseRepo.save(exercise);
+
+        // Add a single participation to that exercise
+        final var exerciseId = exercise.getId();
+
+        var result1 = database.createParticipationSubmissionAndResult(exerciseId, student1, 5.0, 0.0, 60, true);
+        var result2 = database.createParticipationSubmissionAndResult(exerciseId, student2, 5.0, 0.0, 40, true);
```
More vars that at least I don't really like...

FuchsDominik replied:
In this case i would say that the variable names describe the variable well enough. `Result result1` somehow sounds duplicated, doesn't it?

FuchsDominik replied:
Can be discussed in the next meeting

JohannesStoehr commented on src/test/javascript/spec/component/course/detail/course-detail.component.spec.ts:
```diff
@@ -25,23 +24,51 @@ import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
     const course = { id: 123, title: 'Course Title', isAtLeastInstructor: true, endDate: moment().subtract(5, 'minutes'), courseArchivePath: 'some-path' };
+    const dtoMock = {
+        course,
+        numberOfStudentsInCourse: 100,
+        numberOfTeachingAssistantsInCourse: 5,
+        numberOfInstructorsInCourse: 10,
```
The editor is missing

FuchsDominik replied:
Thanks, added in 36c3339 :)

JohannesStoehr reviewed (dismissed):
I'm not really good in client code, so I can't say much about that part.
Apart from the small things I mentioned in the comments, the code looks good.

florian-glombik commented on src/main/java/de/tum/in/www1/artemis/web/rest/dto/CourseManagementDetailViewDTO.java:
```diff
@@ -0,0 +1,190 @@
+        this.course = course;
+    }
+
+    public Integer getNumberOfEditorsInCourse() {
+        return numberOfEditorsInCourse;
+    }
+
+    public void setNumberOfEditorsInCourse(Integer numberOfEditorsInCourse) {
+        this.numberOfEditorsInCourse = numberOfEditorsInCourse;
+    }
```
I think it was better if this part was moved to line 62, directly after `setNumberOfTeachingAssistantsInCourse`

FuchsDominik replied:
Ok 👍

ivohashamov reviewed (dismissed):
Tested on TS2, works as expected

PatZintel reviewed (approved):
didnt find any issues on ts1, even when deleting the for testing created exercise the statistics went back to how they were before

KiraNickel reviewed (approved):
Tested on ts2, everything worked as described 👍🏼

florian-glombik reviewed (changes requested):
Tested it on ts1 and ts3:
- The _Lectures_ button does not work for Students and Editors, I have seen another PR regarding this topic, should be allowed for editors I guess https://github.com/ls1intum/Artemis/pull/3341
![image](https://user-images.githubusercontent.com/63976129/116931967-fc527680-ac61-11eb-8c53-12850b94db88.png)
- Not a big issue, but a margin to the right side of the screen for tutors and editors might be nicer than the _Q&A_ button directly on the right side
![image](https://user-images.githubusercontent.com/63976129/116932384-797deb80-ac62-11eb-9858-47aa10052418.png)

Besides that, it looked good to me

iozbeyli commented on src/test/javascript/spec/component/course/detail/course-detail-bar-chart.component.spec.ts:
```diff
@@ -0,0 +1,87 @@
+
+    it('should initialize', () => {
+        const graphData: number[] = [];
+        const spy = spyOn(service, 'getStatisticsData');
+        for (let i = 0; i < 4; i++) {
+            graphData[i] = 40 + 2 * i;
+        }
+        spy.and.returnValue(of(graphData));
```
What happens to this, you are not using graphData later, not checking if the service method has been called?

FuchsDominik replied:
I use the `graphData` for the service call and i check the service call by checking the values in `component.data`. Since we initialize `data` in percentage and we said to have 50 people in the course, the expected value is double the `graphData` values

FuchsDominik commented on src/test/javascript/spec/component/course/detail/course-detail.component.spec.ts:
```diff
@@ -66,64 +96,35 @@ describe('Course Management Detail Component', () => {
         sinon.restore();
     });
 
-    describe('OnInit for course that has an archive', () => {
-        it('Should call registerChangeInCourses on init', () => {
-            const registerSpy = sinon.spy(comp, 'registerChangeInCourses');
-
-            fixture.detectChanges();
-            comp.ngOnInit();
-            expect(comp.course).to.deep.equal(course);
-            expect(registerSpy).to.have.been.called;
-        });
-    });
```
I think i still have this functionality included in my tests (`Should call registerChangeInCourses on init`). For the registration, we removed the possibility to add yourself as an student when you are an instructor, so this register test is not needed anymore

FuchsDominik commented:
@florian-glombik Thanks for the testing, students should not be able to access the lectures at it is the course management, but you are right that the editors should be able to. Fixed that in 1e80de8 . Also added margin to the Q&A button so it is not squeezed to the page border

ArtemTsurikov reviewed (dismissed):
Looks good!

florian-glombik reviewed (dismissed):
Tested it again on ts1, worked as expected, the issues mentioned in https://github.com/ls1intum/Artemis/pull/3259#pullrequestreview-650719267 are resolved 👍

ivohashamov reviewed (dismissed):
Code changes make sense, as far as I can tell. Just a small remark regarding the use of `!important`. Wouldn't it be possible to increase the specificity of the selectors instead?

MDK5 commented on src/main/webapp/app/course/manage/detail/course-detail-bar-chart.component.ts:
```diff
@@ -0,0 +1,164 @@
+
+    constructor(private service: CourseManagementService, private translateService: TranslateService) {}
+
+    ngOnChanges() {
```
ngOnChanges() and reloadChart() are basically the same code, is it necessary to keep them apart?

MDK5 commented on src/test/java/de/tum/in/www1/artemis/util/CourseTestService.java:
```diff
@@ -1552,4 +1552,147 @@ public void testGetExerciseStatsForCourseOverviewWithPastExercises() throws Exce
         assertThat(statisticsOptional.isEmpty()).isTrue();
     }
+
+    // Test
```
@Test seems to be missing

FrankeLukas replied:
The methods in this file are not annotated with `@Test` since they are helper methods called by the GitlabJenkins and BambooJira tests.

FuchsDominik replied:
Why do we need such an annotation? I don't see any usage of this in this test class

MDK5 reviewed (changes requested):
![detailview](https://user-images.githubusercontent.com/61357727/117060467-3fbfea00-ad21-11eb-92ca-5034518162ad.png)
tested on ts3, detail view and scores absolutely don't match. It might be caused by the weird setup on the test server, I can't tell (course is 3494 on ts3) (I added another submission after taking the screen, but the numbers still dont match)

Code looks good, I only found some small details. Codacy also has a few suggestions that I think are worth looking into.

FuchsDominik commented on src/main/webapp/app/course/manage/detail/course-detail-bar-chart.component.scss:
```diff
@@ -0,0 +1,6 @@
+@media screen and (max-width: 767px) {
+    .course-detail-bar-chart {
+        width: 80% !important;
+        height: 80% !important;
```
I tried using only `width: 80%` but it somehow did not adjust the width as told. `!important` did the trick. Frankly I am not quite sure why `!important` is needed here, maybe has something to do with the canvas automatically setting the width.

MDK5 commented:
> 
> 
> @MDK5 regarding your testing findings, i looked up your course and it seems pretty fine to me. Your course does not have any assessments, no complaints and no more feedback requests. Furthermore, if you look at the participant score where we also display the average score, the average score is indeed the same is in the detail view. (https://vmbruegge60.in.tum.de/course-management/3494/participant-scores, on top is says average score: 12,5%, same as in the detail view for me. Also explained in the testing steps) I did not sum up all the scores listed in the Scores page, but I assume we already display the correct average score in the Participants score page

I see. In that case I wonder why the participant scores show completely different values as the scores page, but that's of no concern for this PR. Sorry for my confusion.

FuchsDominik commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -1386,4 +1392,91 @@ private boolean checkIfUserIsMemberOfCourseOrganizations(User user, Course cours
+        long numberOfSubmissions = submissionRepository.countByCourseIdSubmittedBeforeDueDate(courseId)
+                + programmingExerciseRepository.countLegalSubmissionsByCourseIdSubmitted(courseId);
+        dto.setCurrentMaxAssessments(numberOfSubmissions);
+        if (numberOfSubmissions > 0) {
```
Problem here is that in each check, we set a different variable in the dto. What i can do is extract the whole part like assessment data fetching and assigning into a helper class 👍

MDK5 replied:
I was only suggesting a small helper that gets the number and totalNumber of each type to calculate the percentage and then only replaces the whole if-else by a single call, I totally understand the other parts cannot really be unified.

cremertim reviewed (approved):
Retested on ts 1 . Everything worked as described 👍🏼

MDK5 reviewed (approved):
Tested on t3, works as described; Code also looks good to me

MDK5 reviewed (approved):
Re-Approve

florian-glombik reviewed (approved):
Re-Approve, had a short look at it on ts3 and the changes since the last review - looked fine

JohannesStoehr reviewed (commented):
Tested on ts3, Code looks good as well 👍

JohannesStoehr reviewed (approved):
Tested on ts3, Code looks good as well 👍

ArtemTsurikov reviewed (approved):
Re-Approve. Tested on ts1. Looks good!
